kernel/compareConstr.ml

```diff
@@ -0,0 +1,195 @@
+    | CUMUL -> comp_cst
+
+let convert_inductives_gen cmp_instances cmp_cumul cv_pb (mind,ind) nargs u1 u2 s =
+  match mind.Declarations.mind_universes with
+  | Declarations.Monomorphic_ind _ ->
+    assert (Univ.Instance.length u1 = 0 && Univ.Instance.length u2 = 0);
+    s
+  | Declarations.Polymorphic_ind _ ->
+    cmp_instances u1 u2 s
+  | Declarations.Cumulative_ind cumi ->
+    let num_param_arity =
+      mind.Declarations.mind_nparams +
+      mind.Declarations.mind_packets.(ind).Declarations.mind_nrealargs
+    in
+    if not (Int.equal num_param_arity nargs) then
+      cmp_instances u1 u2 s
+    else
+      let csts = get_cumulativity_constraints cv_pb cumi u1 u2 in
+      cmp_cumul csts s
+
+let convert_constructors_gen cmp_instances cmp_cumul (mind, ind, cns) nargs u1 u2 s =
+  match mind.Declarations.mind_universes with
+  | Declarations.Monomorphic_ind _ ->
+    assert (Univ.Instance.length u1 = 0 && Univ.Instance.length u2 = 0);
+    s
+  | Declarations.Polymorphic_ind _ ->
+    cmp_instances u1 u2 s
+  | Declarations.Cumulative_ind cumi ->
+    let num_cnstr_args =
+      let nparamsctxt =
+        mind.Declarations.mind_nparams +
+        mind.Declarations.mind_packets.(ind).Declarations.mind_nrealargs
+        (* Context.Rel.length mind.Declarations.mind_params_ctxt *) in
+      nparamsctxt + mind.Declarations.mind_packets.(ind).Declarations.mind_consnrealargs.(cns - 1)
+    in
+    if not (Int.equal num_cnstr_args nargs) then
+      cmp_instances u1 u2 s
+    else
+      let csts = get_cumulativity_constraints CONV cumi u1 u2 in
+      cmp_cumul csts s
+
+(* When using compare_head_gen_with we need to return a bool.
+   The argument [s] passed to convert_foo_gen is the value for the monomorphic case, ie true.
+ *)
+let check_globref_univs env univs cv_pb ref nargs u u' =
+  let cmp_instances u u' _ = UGraph.check_eq_instances univs u u' in
+  let cmp_cumul csts _ = UGraph.check_constraints csts univs in
+  match ref with
+  | VarRef _ -> assert false
+  | ConstRef _ ->
+    UGraph.check_eq_instances univs u u'
+  | IndRef ind ->
+    let mind = Environ.lookup_mind (fst ind) env in
+    convert_inductives_gen cmp_instances cmp_cumul cv_pb (mind,snd ind) nargs u u' true
+  | ConstructRef ((mi,ind),ctor) ->
+    let mind = Environ.lookup_mind mi env in
+    convert_constructors_gen cmp_instances cmp_cumul (mind, ind, ctor) nargs u u' true
+
+let infer_globref_univs env univs cv_pb cstsref ref nargs u u' =
+  let cmp_instances u u' _ =
+    if UGraph.check_eq_instances univs u u' then true
+    else begin cstsref := enforce_eq_instances u u' !cstsref; true end
+  in
+  let cmp_cumul csts _ =
+    if UGraph.check_constraints csts univs then true
+    else begin cstsref := Constraint.union csts !cstsref; true end
+  in
+  match ref with
+  | VarRef _ -> assert false
+  | ConstRef _ ->
+    (* If this check fails we don't add constraints as reducing could
+       give a better result. *)
+    UGraph.check_eq_instances univs u u'
```

Comment:
Here we could test if the constant is opaque, in which case adding equality constraints should be OK.

Reply:
Indeed.